network/validator/pubsub/authorized_sender_validator.go

```diff
@@ -15,65 +16,74 @@ import (
 	"github.com/onflow/flow-go/model/flow"
 )
 
-// AuthorizedSenderValidator using the getIdentity func will check if the role of the sender
-// is part of the authorized roles list for the channel being communicated on. A node is considered
-// to be authorized to send a message if the following are true.
-// 1. The node is staked.
-// 2. The node is not ejected.
-// 3. The message type is a known message type (can be decoded with network codec).
-// 4. The message is authorized to be sent on channel.
-// 4. The sender role is authorized to send message channel.
-// 5. The sender role is authorized to participate on channel.
+var (
+	ErrUnauthorizedSender = errors.New("validation failed: sender is not authorized to send this message type")
+	ErrSenderEjected      = errors.New("validation failed: sender is an ejected node")
+	ErrUnknownMessageType = errors.New("validation failed: failed to get message auth config")
+	ErrIdentityUnverified = errors.New("validation failed: could not verify identity of sender")
+)
+
+// AuthorizedSenderValidator returns a MessageValidator that will check if the sender of a message is authorized to send the message.
+// The MessageValidator returned will use the getIdentity to get the flow identity for the sender, asserting that the sender is a staked node.
+// If the sender is an unstaked node the message is rejected. IsAuthorizedSender is used to perform further message validation. If validation
+// fails the message is rejected, if the validation error is an expected error slashing data is collected before the message is rejected.
 func AuthorizedSenderValidator(log zerolog.Logger, channel network.Channel, getIdentity func(peer.ID) (*flow.Identity, bool)) MessageValidator {
-	log = log.With().
+	slashingViolationsConsumer := network.NewSlashingViolationsConsumer(log.With().
 		Str("component", "authorized_sender_validator").
 		Str("network_channel", channel.String()).
-		Logger()
+		Logger())
 
 	return func(ctx context.Context, from peer.ID, msg interface{}) pubsub.ValidationResult {
 		identity, ok := getIdentity(from)
 		if !ok {
-			log.Warn().Str("peer_id", from.String()).Msg("could not verify identity of sender")
+			log.Warn().Err(ErrIdentityUnverified).Str("peer_id", from.String()).Msg("rejecting message")
 			return pubsub.ValidationReject
 		}
 
-		// redundant check if the node is ejected so that we can fail fast before decoding
-		if identity.Ejected {
-			log.Warn().
-				Err(fmt.Errorf("sender %s is an ejected node", identity.NodeID)).
-				Str("peer_id", from.String()).
-				Str("role", identity.Role.String()).
-				Str("node_id", identity.NodeID.String()).
-				Msg("rejecting message")
+		msgType, err := IsAuthorizedSender(identity, channel, msg)
+		if errors.Is(err, ErrUnauthorizedSender) {
```

Comment:
Using errors.Is is more preferable to the switch statement because it will unwrap the error and check if it matches the target err. The switch statement would only check the type of the error 

```
switch err.(type): 
case *Err. . . .
case *Err. . .
```

Also the pattern of using **_errors.New & errors.Is_** is prevalent in the code base when checking sentinel errors.

Reply:
I mean

```
switch {
case errors.Is(err, ErrUnauthorizedSender):
...
case errors.Is(err, ErrUnknownMessageType):
...
case err != nil:
...
}
```